wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,664 @@
+        }
+
+-- Same remark from 'Arbitrary Entropy' applies here.
+instance Arbitrary Mnemonic where
+    arbitrary =
+        entropyToMnemonic <$> arbitrary
+
+instance FromJSON Mnemonic where
+    parseJSON =
+        parseJSON >=> (eitherToParser . mkMnemonic)
+      where
+        eitherToParser :: Either Text Mnemonic -> Parser Mnemonic
+        eitherToParser =
+            either (fail . show) pure
+
+instance ToJSON Mnemonic where
+    toJSON =
+        toJSON . map snd . getWords
+
+instance ToSchema Mnemonic where
+    declareNamedSchema _ = do
+        NamedSchema _ schema <- declareNamedSchema (Proxy @[Text])
+        return $ NamedSchema (Just "Mnemonic") schema
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy :: ByteString -> Either Text Entropy
+mkEntropy bytes = do
+    when (BS.length bytes <= 0) $
+        Left "Entropy must be a non-zero sequence of bytes"
+    when (BS.length bytes > 64) $
+        Left "Maximum entropy is 64 bytes (512 bits)"
+    when (BS.length bytes `rem` 4 /= 0) $
+        Left "Entropy must be a multiple of 4 bytes"
+    Right (Entropy bytes)
+
+
+-- | Smart-constructor for the Mnemonic
+mkMnemonic :: [Text] -> Either Text Mnemonic
+mkMnemonic wordsm = do
+    when (length wordsm <= 0) $
+        Left "Mnemonic must be a non-empty list of words"
+    when (length wordsm > 48) $
+        Left "Mnemonic must have a maximum number of 48 words"
+    when (length wordsm `rem` 3 /= 0) $
+        Left "Mnemonic must be a multiple of 3 words"
+    when (isJust $ Text.find (not . isAscii) (unwords wordsm)) $
+        Left "Mnemonic must use only ASCII characters"
+    when (wordsm == map snd (getWords def)) $
+        Left "Forbidden Mnemonic: an example Mnemonic has been submitted. Please generate a fresh and private Mnemonic from a trusted source"
+    indices <- findIndices enWordList wordsm
+    let bits = concatMap word11ToBits indices
+    let (entropyBits, checksum) = splitAt ((length bits `quot` 33) * 32) bits
+    entropy <- mkEntropy (bitsToBytes entropyBits)
+    when (checksum /= calcChecksum entropy) $
+        Left "mnemonic checksum failed: cannot convert mnemonic to entropy"
+    pure $ Mnemonic
+        { getEntropy = entropy
+        , getWords   = zip indices wordsm
+        }
+  where
+    findIndices :: [Text] -> [Text] -> Either Text [Int]
+    findIndices es targets =
+        case mapM (\e -> elemIndex e es) targets of
+        Nothing ->
+            Left "Unknown Mnemonic word(s)"
+        Just is ->
+            pure is
+
+
+-- | An accessor because we don't want to expose the data-type constructor
+entropyToByteString :: Entropy -> ByteString
+entropyToByteString (Entropy e) =
+    e
+
+
+-- | Convert a mnemonic to a seed that can be used to initiate a HD wallet.
+-- Note that our current implementation deviates from BIP-39 as:
+--
+--    - We do not use the password to produce the seed
+--    - We rely on a fast blake2b hashing function rather than a slow PKBDF2
```

Comment:
From my discussion with @vincenthz on the matter:

> The bip39 uses a password to actually generate its root key, something we don't do on our side that's the reason to use pbkdf2, since the password is not enforced to be particularly anything good more generally every time you accept a user supplied password to do something cryptographic, to prevent having terrible entropy and really fast brute forcing, you go through a layer of slowhashing (pbkdf2, scrypt, bcrypt, argon2).

So, because we don't actually use the user passphrase to compute the seed, there's _no need_ for slow hashing. The entropy from the mnemonic is already big enough to cope with brute-force.

Reply:
Beside, I don't know why it was done this way.